SSA/Projects/InstCombine/Tactic.lean

```diff
@@ -12,26 +12,46 @@ open MLIR AST
 open Std (BitVec)
 open Ctxt
 
+-- TODO: Do we call this dialect the "Alive" dialect (as hinted by the `simp_alive_peephole` name),
+--       or the "InstCombine" dialect (as hinted by the `InstCombine` namespace that everything
+--       lives in), or the "LLVM" dialect (as we do in the paper, and the simp-sets,
+--       and the `LLVM` namespace)?
+
 open MLIR AST in
 /--
-- We first simplify `Com.refinement` to see the context `Γv`.
-- We `simp_peephole Γv` to simplify context accesses by variables.
-- We simplify the translation overhead.
-- Then we introduce variables, `cases` on the variables to eliminate the `none` cases.
-- We cannot leave it at this state, since then the variables will be inaccessible.
-- So, we revert the variables for the user to re-introduce them as they see fit.
--/
+`simp_alive_peephole` extends `simp_peephole` to simplify goals about refinement of `LLVM`
+programs into statements about just bitvectors.
+
+To wit, the tactic expects a goal of the form: `Com.Refinement com₁ com₂`
+That is, goals of of the form `Com.refine, com₁.denote Γv ⊑ com₂.denote Γv `,
+where `com₁` and `com₂` are programs in the `LLVM` dialect. -/
 macro "simp_alive_peephole" : tactic =>
   `(tactic|
       (
+        /- Unfold the meaning of refinement, to access the valuation -/
         dsimp only [Com.Refinement]
         intros Γv
+
+        /- Simplify away the core framework -/
         simp_peephole [InstCombine.Op.denote] at Γv
+
+        /- Simplify away the `InstCombine` specific semantics. -/
         simp (config := {failIfUnchanged := false}) only [
             BitVec.Refinement, bind, Option.bind, pure,
             simp_llvm,
             BitVec.bitvec_minus_one
           ]
+
+        /-  At this point, we have should a goal of the form:
+              `∀ (x₁ : Option (BitVec _)) ... (xₙ : Option (BitVec _)), ...`
+            For some unknown number of variables `n` (but assumed to be `n ≤ 5`,
+            as per the hack in `simp_peephole`).
+
+            However, the `none` (i.e., poison) cases are generally trivial.
+            Thus we attempt to introduce some variables (using `try` because the intro might fail
+            if we have less universal quantifiers than `intro`s), case split on them, and simp
+            through the monadic bind on `Option` (in the generally true assumption that the `none`
+            case becomes trivial and is closed by `simp`). -/
```

Comment:
Would it make sense to outline this into a separate tactic and then document this tactic?

Reply:
I've extracted it into a new, local `simp_alive_case_bash` tactic